### PR TITLE
Proposed 2.5 - restore docker-extract whiteout handling

### PR DESCRIPTION
**Pull request by dtrudg**
**Description of the Pull Request (PR):**

This PR restores the docker-extract whiteout handling that was introduced in 2.4.3, but removed in 2.4.4. The PR ensures that when docker layers are extracted, whiteout .wh.* files are correctly handled, resulting in a correct container fs.

At runtime, this PR adds the ability for extraction to fall-back to the previous approach (using tar) if libarchive is not available.

- libarchive (with headers) is *required* at build time\*
- libarchive at runtime enables correct whiteout handling
- missing libarchive at runtime falls back to incorrect extraction, no whiteout handling

\*Note - it is possible to run 'make -i install' on a machine without libarchive headers, if 'make' was run at a location with these*

Thanks to @DrDaveD for providing the commit which removed the unintentional libarchive linking to other binaries that was observed in 2.4.3 on some distributions.

Changes to the following files in this commit are *not* new, and have been reviewed in previous PRs:

- .gitignore
- .travis.yml
- .travis/rpmbuild_test
- bin/singularity.in
- debian/control
- singularity.spec.in
- src/Makefile.am
- src/docker-extract.c
- tests/21-build_docker.sh

Changes to the following files in this commit *are* new, and have not yet been reviewed:

- CHANGELOG.md
- configure.ac
- libexec/bootstrap-scripts/deffile-driver-docker.sh
- libexec/bootstrap-scripts/main-dockerhub.sh
- libexec/cli/build.exec
- libexec/handlers/image-docker.sh

**Checkoff for all PRs:**

- [X] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [X] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [X] I have tested this PR locally with a `make test`
- [X] This PR is NOT against the project's master branch
- [X] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [X] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

File: src/docker-extract.c [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+        singularity_message(ERROR, "Error with pathname too long\n");
+        ABORT(255);
+    }
+
+    // Target may not exist - that's ok
+    retval = 0;
+
+    if (is_dir(buf) == 0) {
+        retval = s_rmdir(buf);
+    } else if (is_file(buf) == 0) {
```
Comment (cclerget) [inline src/docker-extract.c]: @dctrud symlinks seems not concerned by this check but need to be removed too

Comment (dtrudg) [reply src/docker-extract.c]: @cclerget - thanks for spotting that. Added an `|| is_link(buf) == 0` on the else if

**Comment by jakirkham** [comment]
What about statically linking `libarchive`? Would that work or is that somehow problematic?

**Comment by dtrudg** [comment]
Unfortunately the .a isn't available in easy distro packages on all common distros.

**Comment by jakirkham** [comment]
That could be solved by vendoring `libarchive`. Not ideal, but it's an option.

**Comment by dtrudg** [comment]
@jakirkham - yes it was one of the things we discussed but we are trying to get this out sooner rather than later - and there are complexities integrating into the build and requiring us to follow libarchive security updates in our tree etc.

At this point in time, lack of a libarchive .so at runtime will drop back to old extraction behaviour - which hopefully eases things for people who really can't push libarchive out to cluster nodes easily.

**Comment by jakirkham** [comment]
Ok, sorry, took the removed release label above as this wasn't going to make it into 2.5.

**Comment by dtrudg** [comment]
Ahh - no... I just switched from label to milestone for tracking. It's in the 2.5 milestone :-)

**Comment by jakirkham** [comment]
Great! Thanks for trying to get it in. :)

Is there a particular version range of libarchive that we need? Was going to nudge our cluster admins if we don't have it already (we probably do).

**Review by cclerget — APPROVED** [review]
Looks good to me

**Review by GodloveD — APPROVED** [review]
I've been tinkering with this for a while now and it all seems good to me.

**Pull request merged**